// leevueplugin/index.js
import LeeAlert from './src/LeeAlert.vue'
import LeeBadge from './src/LeeBadge.vue'
import LeeButton from './src/LeeButton.vue'
import LeeCollapse from './src/LeeCollapse.vue'
import LeeCollapseItem from './src/LeeCollapseItem.vue'
import LeeInputNumber from './src/LeeInputNumber.vue'
import LeeNotification from './src/LeeNotification.vue'
import LeeNotificationItem from './src/LeeNotificationItem.vue'
import LeePagination from './src/LeePagination.vue'
import LeePasswordInput from './src/LeePasswordInput.vue'
import LeePay from './src/LeePay.vue'
import LeeRadio from './src/LeeRadio.vue'
import LeeRadioGroup from './src/LeeRadioGroup.vue'
import LeeSwitch from './src/LeeSwitch.vue'
import LeeTimeline from './src/LeeTimeline.vue'
import LeeTimelineItem from './src/LeeTimelineItem.vue'

LeeAlert.install = function(Vue) {
    Vue.component(LeeAlert.name, LeeAlert)
}
LeeBadge.install = function(Vue) {
    Vue.component(LeeBadge.name, LeeBadge)
}
LeeButton.install = function(Vue) {
    Vue.component(LeeButton.name, LeeButton)
}
LeeCollapse.install = function(Vue) {
    Vue.component(LeeCollapse.name, LeeCollapse)
}
LeeCollapseItem.install = function(Vue) {
    Vue.component(LeeCollapseItem.name, LeeCollapseItem)
}
LeeInputNumber.install = function(Vue) {
    Vue.component(LeeInputNumber.name, LeeInputNumber)
}
LeeNotification.install = function(Vue) {
    Vue.component(LeeNotification.name, LeeNotification)
}
LeeNotificationItem.install = function(Vue) {
    Vue.component(LeeNotificationItem.name, LeeNotificationItem)
}
LeePagination.install = function(Vue) {
    Vue.component(LeePagination.name, LeePagination)
}
LeePasswordInput.install = function(Vue) {
    Vue.component(LeePasswordInput.name, LeePasswordInput)
}
LeePay.install = function(Vue) {
    Vue.component(LeePay.name, LeePay)
}
LeeRadio.install = function(Vue) {
    Vue.component(LeeRadio.name, LeeRadio)
}
LeeRadioGroup.install = function(Vue) {
    Vue.component(LeeRadioGroup.name, LeeRadioGroup)
}
LeeSwitch.install = function(Vue) {
    Vue.component(LeeSwitch.name, LeeSwitch)
}
LeeTimeline.install = function(Vue) {
    Vue.component(LeeTimeline.name, LeeTimeline)
}
LeeTimelineItem.install = function(Vue) {
    Vue.component(LeeTimelineItem.name, LeeTimelineItem)
}

export {LeeAlert,LeeBadge,LeeButton,LeeCollapse,LeeCollapseItem,LeeInputNumber,LeeNotification,LeeNotificationItem,LeePagination,LeePasswordInput,LeePay,LeeRadio,LeeRadioGroup,LeeSwitch,LeeTimeline,LeeTimelineItem}
